pitchtypes/spelled.py

```diff
@@ -1,34 +1,32 @@
 #  Copyright (c) 2022 Robert Lieck, Christoph Finkensiep
 
 import numbers
-import re
 import abc
 import functools
 
 import numpy as np
 
 from pitchtypes.basetypes import AbstractBase, Pitch, Interval, Diatonic, Chromatic
+from pitchtypes import utils
+
 
 @functools.total_ordering
 class Spelled(AbstractBase):
     """
     A common base class for spelled pitch and interval types.
     See below for a set of common operations.
     """
-    _pitch_regex = re.compile("^(?P<class>[A-G])(?P<modifiers>(b*)|(#*))(?P<octave>(-?[0-9]+)?)$")
```

Comment:
Since we are also parsing pitches and intervals in enharmonic.py and generic.py I feel like it's better to either put these methods in utils.py or in the superclasses Pitch and Interval instead of calling Spelled.parse_pitch and Spelled.parse_interval. I chose the latter, but let me know if there's a better approach.

Reply:
This particular regex is only meaningful for spelled pitches. Enharmonic pitches are notated differently (usually as Midi numbers), so there is no reason to use this regex there. I'm not sure what the generic types are supposed to do, maybe you can explain that briefly?

Reply:
I considered that the user is allowed, for example, to create an enharmonic pitch from a spelled pitch and that it will be converted into the midi representation in the constructor, e.g. "G#4" -> 68. This is just to give the user some freedom, but we can remove this.

Reply:
We had some discussion on Mattermost with Johannes about implicit conversions but we haven't really agreed on how to do this yet. Currently, the preferred way to do something like this is to do the conversion explicitly, i.e. parse "G#4" with the appropriate class (SpelledPitch) and then convert it to an enharmonic pitch.

If you want to allow the user to do this in one step, you can still avoid code duplication by using the constructor of SpelledPitch to do the parsing. This way the code for parsing spelled pitches stays in the right location.

Reply:
I did some testing and in my opinion it is best to keep the parsers (parse_pitch and parse_interval) in their corresponding superclasses Pitch and Interval. The initial code, which involved calling the parse methods from the Spelled class, is complex for little reason, as there exist use cases where we can create Generic and Enharmonic types from strings, without Spelled needing to be implicated. It’s more logical to handle these methods at the superclass level rather than having classes on the same level call methods from each other.

The other viable solution from an architectural standpoint, which is creating an instance of Spelled and converting it to Enharmonic or Generic using the converters, is really inefficient. I used timeit to measure the time taken to create some Enharmonic objects and this solution is almost 7 times slower than the other 2.

```
import timeit

def creating_EPs():
    ep = EnharmonicPitch("C#4")

execution_time = timeit.timeit(creating_EPs, number=1000000)
print(f"Execution time: {execution_time} seconds")

# 5.36 seconds for parsers in their corresponding superclasses Pitch and Interval
# 34.04 seconds for creating Spelled object and converting to Enharmonic
# 6.58 seconds for parser in Spelled and call it from Enharmonic
```

Reply:
I'm not a big fan of this kind of implicit conversion you have in `EnharmonicPitch("C#4")`. I think it would be much cleaner to have the user explicitly make the conversion by first parsing a spelled pitch (which C#4 *is*) and then convert it to enharmonic to clearly show the intentions of the code and make explicit what is happening. So I'd be in favor of your second solution.

However, I can also see the use case where you might want to parse a dataset where the exact notation format is not always known or mixed. I'd say the "right" solution would be to do some data cleaning first (you should ideally know what data you are using, in a scientific context), but I can understand that a bit convenience might be nice here. However, in that case, the question is if we can use the same regex to parse stuff in that case, or if we need to be able to distinguish between different string formats at once (e.g. "G#4" vs "68" vs "p64" vs "G4" (generic) etc.). I'm a bit torn on this, I think I'd prefer clean and predictable solutions over convenience here.

I don't think performance should be the main concern here. If you spend 35s on 1M notes, that's not a big deal, most other operations data operations are going to be more expensive. The better solution would be to have array variants for all types so that we get cheaper conversion for larger methods. Also, it's a bit strange that the extra 30s come just from the conversion (a simple arithmetic operation) when the parser (matching a regex, interpreting the result and the same kinds of arithmetics) only needs 5s, that just doesn't make sense.